### ISSUE TITLE
Failed user registration redirects to /users which isn't a valid get url

### ISSUE THREAD

**Issue by tae8838** [plataformatec/devise#4573]
After clicking `Sign up`, the url changes to `/users` because we post to `/resource`. If failed, we do `respond_with resource` without redirecting. If refreshed, `No route matches [GET] "/users"` appears.

Could we have failed sign up redirect to the same page?

**Comment by tegon** [comment]
As far as I know, this is the default of a Rails scaffold application.
But I don't see a problem in changing this behavior.
WDYT @feliperenan @rafaelfranca?

**Comment by iorme1** [comment]
I will be working on a PR for this issue within the next few days.

**Comment by iorme1** [comment]
Unfortunately, redirecting to the same page here loses the "devise_error_messages!" method call in the view because the resource.errors is empty after redirect, and the devise_error_messages! call checks to see if 'resource.errors.empty?' before returning the sentence about the errors and error messages.  You can redirect to the same page and keep the error messages themselves, but it won't display the sentence/error display like it does currently through the devise_error_messages! call in the view. I can't seem to figure out a way around this. I would really like to address this and make a PR, so any help would be much appreciated!

**Comment by iorme1** [comment]
@tegon , just making sure you see my last message on this issue. Trying to clarify how you want this issue solved exactly.

**Comment by lancecarlson** [comment]
@iorme1 It's kind of a crappy hack, but you could POST to /users/sign_up instead of /users. The alternative would be to hack up the way devise/rails supports validation errors and store it in the session.

**Comment by tegon** [comment]
@iorme1 Sorry for taking so long, I was on vacation and didn't see this. I don't have a better solution right now, I'll have to think better about it.

**Comment by JuanJ0424** [comment]
I also dealing with this problem, and have a similiar one. Sorry for my little experience. I'm adding functionality to the sign_up action and I cannot do the  repond_with resource from my controller, cause it raises a NoMethodError for "users_url", but this is done without any problem by the super method.

**Comment by sarunw** [comment]
What is the solution for this?

**Comment by tegon** [comment]
Sorry, I haven't take a look at this yet. This is the default behavior we get when scaffolding a Rails CRUD, so I'm not sure what path to follow here.

**Comment by connorshea** [comment]
I had this problem when using the registrations#edit form in a custom settings view.

I ended up replacing the `update` path in my registrations controller with, essentially, [the code from the devise repo](https://github.com/plataformatec/devise/blob/07f2712a22aa05b8da61c85307b80a3bd2ed6c4c/app/controllers/devise/registrations_controller.rb#L43-L62) instead of calling `super`.

Before:

```ruby
# app/controllers/users/registrations_controller.rb

# PUT /resource
def update
  skip_authorization
  super
end
```

After:

```ruby
# app/controllers/users/registrations_controller.rb

# PUT /resource
def update
  skip_authorization
  self.resource = resource_class.to_adapter.get!(send(:"current_#{resource_name}").to_key)
  prev_unconfirmed_email = resource.unconfirmed_email if resource.respond_to?(:unconfirmed_email)

  resource_updated = update_resource(resource, account_update_params)
  yield resource if block_given?
  if resource_updated
    set_flash_message_for_update(resource, prev_unconfirmed_email)
    bypass_sign_in resource, scope: resource_name if sign_in_after_change_password?

    respond_with resource, location: after_update_path_for(resource)
  else
    clean_up_passwords resource
    set_minimum_password_length
    # Redirect to the settings page.
    redirect_to settings_account_path
  end
end
```

The only changes I made were to add `skip_authorization`, which is a Pundit thing that's not really relevant, and changing the `respond_with resource` line to `redirect_to settings_account_path` so it would redirect to the page I wanted.

This works (better than it redirecting to `/users` on error, at least), but it doesn't pass the devise error messages. Any ideas on how I'd go about passing the errors to the redirect?

**Comment by tegon** [comment]
Hi everyone, 

After giving it some thought the only way to achieve the persistence of error messages seems to be to store them in the session, cookies or some of the alike (please let me know if you have any other ideas).
I don't think the complexity of this is worth it adding to the library and since this is the default Rails' resource behavior, we'll leave as it is for now and let each application handle the way it better fits them.

Hope you understand my reasoning. Thanks.

**Comment by jjb** [comment]
What's the TLDR on why only registrations has this problem and not sign in and other forms? (btw I am not experiencing this on password reset, which this comment says also has the problem, I wonder if a fix was put in for that in the meantime, but not for registrations https://github.com/heartcombo/devise/issues/4573#issuecomment-445018377 )

**Comment by jjb** [comment]
This was all I needed to fix my problems, devise 4.8.1. Maybe I'm missing something that I'll discover later.

```ruby
get '/users', to: redirect('/users/sign_up')
```

**Comment by 404UsernameNotFound404** [comment]
There are lots of solution that fix it, but don't keep the error message. Is there any update on this? My version is 4.8.1
